### PR TITLE
Matrices with random expressions

### PR DESCRIPTION
**Pull request by czgdp1807**
#### Brief description of what is fixed or changed
N/A

#### Other comments
ping @Upabjojr @sidhantnagpal 

#### Release Notes

* stats
  * Matrices with random expressions have been added in `sympy.stats`.

### CODE REVIEW

File: sympy/stats/random_matrix_models.py [hunk 1]
```diff
@@ -212,6 +214,45 @@ def joint_eigen_distribution(mat):
...
+
+    """
+    eigenvals, counts = mat.eigenvals().keys(), mat.eigenvals().values()
+    if any(not eigenval.has(RandomSymbol) for eigenval in eigenvals):
+        raise ValueError("Eigen values don't have any random expression, "
+                         "joint distribution cannot be generated.")
+    expanded = []
+    for eigenval, count in zip(eigenvals, counts):
+        l = [eigenval]*count
+        expanded.extend(l)
```
Comment (sidhantnagpal) [inline sympy/stats/random_matrix_models.py]: If you want to do it in a single line:
```py
expanded = [e for ev, cnt in zip(eigenvals, counts) for e in [ev]*cnt]
```

File: sympy/stats/random_matrix_models.py [hunk 2]
```diff
@@ -212,6 +214,45 @@ def joint_eigen_distribution(mat):
     """
     return mat.pspace.model.joint_eigen_distribution()
 
+@dispatch(Matrix)
+def joint_eigen_distribution(mat):
```
Comment (sidhantnagpal) [inline sympy/stats/random_matrix_models.py]: Should this be named `JointEigenDistribution` (even though it is not a class, it still returns an unevaluated object)?

File: sympy/stats/random_matrix_models.py [hunk 3]
```diff
@@ -212,6 +214,45 @@ def joint_eigen_distribution(mat):
     """
     return mat.pspace.model.joint_eigen_distribution()
 
+@dispatch(Matrix)
```
Comment (Upabjojr) [inline sympy/stats/random_matrix_models.py]: are you going to use this function here only? If so, no need to dispatch. Multiple dispatching is only needed if you dispatch over different files.

File: sympy/stats/random_matrix_models.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/stats/random_matrix_models.py]: Ah! I was using unpacking through `*`, that's why it didn't work. Thanks for this idea. I will update.

File: sympy/stats/random_matrix_models.py (hunk 3 above)
Comment (czgdp1807) [reply sympy/stats/random_matrix_models.py]: Yes, I will apply, https://github.com/sympy/sympy/pull/17306#discussion_r309337663, so we will not required to use `dispatch`.

File: sympy/stats/random_matrix_models.py [hunk 4]
```diff
@@ -207,11 +209,49 @@ def joint_eigen_distribution(mat):
...
+    >>> A = [[Normal('A00', 0, 1), Normal('A01', 0, 1)],
+    ... [Normal('A10', 0, 1), Normal('A11', 0, 1)]]
+    >>> JointEigenDistribution(Matrix(A))
+    JointDistributionHandmade(-sqrt(A00**2 - 2*A00*A11 + 4*A01*A10 + A11**2)/2
+    + A00/2 + A11/2, sqrt(A00**2 - 2*A00*A11 + 4*A01*A10 + A11**2)/2 + A00/2 + A11/2)
+
+    """
+    eigen_dict = mat.eigenvals()
+    eigenvals = eigen_dict.keys()
+    sorted(eigenvals, key=lambda val: eigen_dict[val])
```
Comment (czgdp1807) [inline sympy/stats/random_matrix_models.py]: I have tried `eigenvals` canonical by using `sorted`. @sidhantnagpal will the key(multiplicity of the eigen value ) always work?

Comment (sidhantnagpal) [reply sympy/stats/random_matrix_models.py]: You can get the list (with multiplicity) using:
```py
evs = mat.eigenvals(multiple=True)
```
and return the result like `JointDistributionHandmade(*evs)`.

Comment (czgdp1807) [reply sympy/stats/random_matrix_models.py]: Will it always be the same or the order will be random?

Comment (sidhantnagpal) [reply sympy/stats/random_matrix_models.py]: `eigenvals` internally calls [`roots`](https://docs.sympy.org/latest/modules/polys/reference.html#sympy.polys.polyroots.roots) which will make same values appear next to each other.

File: sympy/stats/random_matrix_models.py [hunk 5]
```diff
@@ -207,11 +209,47 @@ def joint_eigen_distribution(mat):
     >>> from sympy.stats import GaussianUnitaryEnsemble as GUE
     >>> from sympy.stats import joint_eigen_distribution
     >>> U = GUE('U', 2)
-    >>> joint_eigen_dsitribution(U)
+    >>> joint_eigen_distribution(U)
     Lambda((l[1], l[2]), exp(-l[1]**2 - l[2]**2)*Product(Abs(l[_i] - l[_j])**2, (_j, _i + 1, 2), (_i, 1, 1))/pi)
     """
+    if not isinstance(mat, RandomMatrixSymbol):
+        raise ValueError("%s is not of type, RandomMatrixSymbol.")
```
Comment (Upabjojr) [inline sympy/stats/random_matrix_models.py]: forgot to fill `%s` ?

**Pull request merged**